docs/ql-libraries/dataflow/dataflow.md

```diff
@@ -0,0 +1,472 @@
+
+### Field flow barriers
+
+Consider this field flow example:
+```
+obj.f = source;
+obj.f = safeValue;
+sink(obj.f);
+```
+or the similar case when field flow is used to model collection content:
+```
+obj.add(source);
+obj.clear();
+sink(obj.get(key));
+```
+Clearing a field or content like this should act as a barrier, and this can be
+achieved by marking the relevant `Node, Content` pair as a clear operation in
+the `clearsContent` predicate. A reasonable default implementation for fields
+looks like this:
+```ql
+predicate clearsContent(Node n, Content c) {
+  n = any(PostUpdateNode pun | storeStep(_, c, pun)).getPreUpdateNode()
+}
+```
+However, this relies on the local step relation using the smallest possible
+use-use steps. If local flow is implemented using def-use steps, then
+`clearsContent` might not be easy to use.
+
+## Type pruning
+
+The library supports pruning paths when a sequence of value-preserving steps
+originate in a node with one type, but reaches a node with another and
+incompatible type, thus making the path impossible.
+
+The type system for this is specified with the class `DataFlowType` and the
+compatibility relation `compatibleTypes(DataFlowType t1, DataFlowType t2)`.
+Using a singleton type as `DataFlowType` means that this feature is effectively
+disabled.
+
+It can be useful to use a simpler type system for pruning than whatever type
+system might come with the language, as collections of types that would
+otherwise be equivalent with respect to compatibility can then be represented
+as a single entity (this improves performance). As an example, Java uses erased
+types for this purpose and a single equivalence class for all numeric types.
+
+One also needs to define
+```
+Type Node::getType()
+Type Node::getTypeBound()
+DataFlowType getErasedRepr(Type t)
+string ppReprType(DataFlowType t)
+```
+where `Type` can be a language-specific name for the types native to the
+language.  Of the member predicate `Node::getType()` and `Node::getTypeBound()`
+only the latter is used by the library, but the former is usually nice to have
+if it makes sense for the language. The `getErasedRepr` predicate acts as the
+translation between regular types and the type system used for pruning, the
+shared library will use `getErasedRepr(node.getTypeBound())` to get the
+`DataFlowType` for a node. The `ppReprType` predicate is used for printing a
+type in the labels of `PathNode`s, this can be defined as `none()` if type
+pruning is not used.
+
+Finally, one must define `CastNode` as a subclass of `Node` as those nodes
+where types should be checked. Usually this will be things like explicit casts.
+The shared library will also check types at `ParameterNode`s and `OutNode`s
+without needing to include these in `CastNode`.  It is semantically perfectly
+valid to include all nodes in `CastNode`, but this can hurt performance as it
+will reduce the opportunity for the library to compact several local steps into
+one.
```

Comment:
Maybe state somewhere here that it's also semantically valid for no node to be a `CastNode`. And is it also worth stating that `CastNode`s are a hacky way to break up the big step relation?

Reply:
Indeed, after reading that making all nodes `CastNode`s could hurt performance, Python switched to making no nodes `CastNode`s.